### PR TITLE
transaction TTL integration

### PR DESCRIPTION
**Pull request by eugene-babichenko**
- Update chain-libs version
- Implement block0 configuration for TTL
- Update jcli to make a user able to set TTL for a transaction
- Update tests
- mjolnir is using *dummy* values and is not functional right now
- Show txMaxExpiryEpochs in REST API.

### CODE REVIEW

**Comment by eugene-babichenko** [comment]
The REST API update is held back by some silliness in `chain-impl-mockchain` Ledger API which I am going to try fixing.

**Comment by mzabaluev** [comment]
The audit failure is resolved by #3506.

File: jcli/src/jcli_lib/transaction/set_valid_until.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+use structopt::StructOpt;
+
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub struct SetValidUntil {
+    #[structopt(flatten)]
+    pub common: common::CommonTransaction,
+
+    /// the slot this transaction should be valid until, for example 3.14
+    #[structopt(name = "VALID_UNTIL")]
```
Comment (mzabaluev) [inline jcli/src/jcli_lib/transaction/set_valid_until.rs:12]: What's the renaming for, the kebab-case should work already?
This maybe needs `value_name = "BLOCKDATE"`

File: jcli/src/jcli_lib/transaction/staging.rs [hunk 2]
```diff
@@ -95,6 +98,16 @@ impl Staging {
         })
     }
 
+    pub fn set_validity_time(&mut self, input: interfaces::BlockDate) -> Result<(), Error> {
```
Comment (mzabaluev) [inline jcli/src/jcli_lib/transaction/staging.rs:101]: Name `set_valid_until` for consistency? Or better, `set_expiry_date` everywhere because that's the name used in the library API.

File: jormungandr/src/fragment/pool.rs [hunk 3]
```diff
@@ -264,9 +264,26 @@ impl Pools {
...
+            fragment_ids,
+            FragmentStatus::Rejected {
+                reason: "fragment expired".to_string(),
+            },
+            block_date.into(),
+        );
+    }
 }
 
 fn is_fragment_valid(fragment: &Fragment) -> bool {
```
[the remark this answers is not in the document]
Comment (mzabaluev) [reply jormungandr/src/fragment/pool.rs:289]: This is a quick and cheap check, I think we should.

File: jormungandr/src/fragment/pool.rs [hunk 4]
```diff
@@ -435,15 +453,61 @@ pub(super) mod internal {
...
         max_entries: usize,
     }
 
     impl Pool {
         pub fn new(max_entries: usize) -> Self {
             Pool {
                 entries: IndexedDeqeue::new(),
+                // Using BTreeSet is a nasty hack so that we are able to to efficiently remove items
+                // out of their order in a queue. BinaryHeap does not allow that and even if it did,
+                // the worst case complexity of removal of a non-root element would be O(n).
```
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply jormungandr/src/fragment/pool.rs:509]: I agree. The only problem is that I would need to implement or import one more custom data structure for that and I did not want to.

[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply jormungandr/src/fragment/pool.rs:509]: P.S. I implied using `std::collections`.

File: jormungandr/src/fragment/pool.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (eugene-babichenko) [reply jormungandr/src/fragment/pool.rs:296]: It is relatively easy. Although we would have to either access the blockchain tip every time we insert a fragment into the pool or store the current block date and update it from the outside.

File: testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs [hunk 5]
```diff
@@ -64,6 +65,7 @@ pub fn test_mempool_pool_max_entries_limit() {
         .transaction_to(
             &jormungandr.genesis_block_hash(),
             &jormungandr.fees(),
+            chain_impl_mockchain::block::BlockDate::first(),
```
[the remark this answers is not in the document]
Comment (Zeegomo) [reply testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs:68]: Are we really using chain-wallet-libs here? I think this is using https://github.com/input-output-hk/jormungandr/blob/0c128de718718b4c00f3f67e54763734bb98e5b9/testing/jormungandr-testing-utils/src/wallet/mod.rs#L320, which is just forwards the `BlockDate` to https://github.com/input-output-hk/jormungandr/blob/47c4ba0cf60be9a999a1b02d9e3e3e6d3e8ee5b5/testing/jormungandr-testing-utils/src/testing/fragments/transaction.rs#L25

File: jormungandr/src/fragment/pool.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mzabaluev) [reply jormungandr/src/fragment/pool.rs:296]: This should work in passive nodes, and we don't have a separate time task that would signal slot changes. For this check, the current block date can be derived from blockchain settings and the system time. It should be a rare occurrence that a fragment is received at the edge of its validity window, no need to be perfectly synchronized.

File: testing/jormungandr-integration-tests/src/common/jcli/services/transaction_builder.rs [hunk 6]
```diff
@@ -73,11 +76,13 @@ impl TransactionBuilder {
         sender: &Wallet,
         input_amount: Value,
         receiver: &Wallet,
+        expiry_date: BlockDate,
```
Comment (mzabaluev) [inline testing/jormungandr-integration-tests/src/common/jcli/services/transaction_builder.rs:79]: Why even have this function that triggers a lint, when all it does is to invoke the proper builder API?

**Review by mzabaluev — COMMENTED** [review]
Looks OK except the linter breakage.

File: testing/jormungandr-integration-tests/src/common/jcli/services/transaction_builder.rs (hunk 6 above)
Comment (eugene-babichenko) [reply testing/jormungandr-integration-tests/src/common/jcli/services/transaction_builder.rs:79]: Well, that is a valid question. Especially given that it is used only 2 times.

File: testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (mzabaluev) [reply testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs:68]: I can't see where does the builder increment the value. This looks like the tests would still pass (and expired fragments validate in production?) if the TTL check was not working as expected.

**Comment by eugene-babichenko** [comment]
There is an error seemingly created by copy-pasting things, fixing.

File: testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs [hunk 7]
```diff
@@ -53,6 +53,7 @@ pub fn test_mempool_pool_max_entries_limit() {
...
```
Comment (mzabaluev) [inline testing/jormungandr-integration-tests/src/jormungandr/bft/mempool/v1.rs:56]: How can this transaction be accepted if we have a check for TTL on entry to the mempool, and @eugene-babichenko has updated chain-libs to include https://github.com/input-output-hk/chain-libs/pull/617?

**Review by mzabaluev — APPROVED** [review]
Use of `BlockDate::first()` as a currently valid TTL in many tests highlights the questionable current behavior: the "valid until" date is considered valid inclusively, contrary to the name of the property. Let's go ahead with this and address the edge case behavior in a later change.
